File: JavaScript/135-Candy.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function(ratings) {
    let candy = new Array(ratings.length).fill(1)
    
    for (let i=1; i<ratings.length; i++){
        if (ratings[i] > ratings[i-1]){
            candy[i] = candy[i-1] + 1
        }
    }
    
    let res = candy[candy.length-1];
    
    for (let i=ratings.length-2; i>=0; i--){
        if (ratings[i] > ratings[i+1]){
            candy[i] = Math.max(candy[i], candy[i+1] + 1);
        }
        res += candy[i];
    }
    return res;
};

File: C++/135-Candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int N = ratings.size();
        vector<int> tot(N, 1);

        // increase candies only when rating increases, no operations otherwise 

        for (int i=1; i<N; i++) {
            if (ratings[i] > ratings[i-1]) {
                tot[i] = tot[i-1] + 1;
            }
        }

        for (int i=N-2; i>=0; i--) {
            if (ratings[i] > ratings[i+1]) {
                tot[i] = max(tot[i], tot[i+1] + 1);
            }
        }

        return accumulate(begin(tot), end(tot), 0);
    }
};